docs/extensions/api.md

```diff
@@ -26,34 +28,51 @@ other issues on the [bug tracker].
 
 ## Preprocessors {: #preprocessors }
 
-Preprocessors munge the source text before it is passed into the Markdown
-parser. This is an excellent place to clean up bad syntax, extract things the
-parser may otherwise choke on and perhaps even store it for later retrieval.
+Preprocessors munge the source text before it is passed to the Markdown
+parser. This is an excellent place to clean up bad encodings or to extract 
+portions for later processing that the parser may otherwise choke on.
 
 Preprocessors should inherit from `markdown.preprocessors.Preprocessor` and
-implement a `run` method with one argument `lines`. The `run` method of
-each Preprocessor will be passed the entire source text as a list of Unicode
-strings. Each string will contain one line of text. The `run` method should
-return either that list, or an altered list of Unicode strings.
-
+implement a `run` method, which takes a single parameter `lines`. This parameter is
+the entire source text stored as a list of Unicode string, one per line.  `run` should
+return its processed list of of Unicode strings, one per line.
 
-A pseudo example:
+A simple example:
 
 ```python
 from markdown.preprocessors import Preprocessor
+import re
 
-class MyPreprocessor(Preprocessor):
+class NoRender(Preprocessor):
+    """ Skip any line with words 'NO RENDER' in it. """
     def run(self, lines):
         new_lines = []
         for line in lines:
-            m = MYREGEX.match(line)
+            m = re.search("NO RENDER", line)
             if m:
-                # do stuff
+                pass   # skipping this line
             else:
                 new_lines.append(line)
         return new_lines
 ```
 
+Some preprocessors in the Markdown source tree include:
+
+
+| Name  | kind | Priority |  Description |
+| ----------------------------|-----------|----|-----------------------------------------------
+| [NormalizeWhiteSpace][c1]   | built-in  | 30 | Normalizes whitespace by expanding tabs, fixing `\r` line endings, etc. |
+| [HtmlBlockPreprocessor][c2] | built-in  | 20 | Removes html blocks from the text and stores them for later processing |
+| [ReferencePreprocessor][c3] | built-in  | 10 | Removes reference definitions from text and stores for later processing |
+| [MetaPreprocessor][c4]      | extension | 27 | Strips and records meta data at top of documents |
+| [FootnotesPreprocessor][c5] | extension | 15 | Removes footnote blocks from the text and stores them for later processing |
```

Comment:
This is interesting. I don't think I had considered doing this. The problem is that the links to the code are not for a specific blob. Therefore, as changes happen in the future, the line numbers will become wrong. At the same time, updating the blob id is not something we want to do either.

Perhaps instead we should be using [tomchristie/mkautodoc][1] to pull in the documentation from the source code for the Python objects. To be clear, we still want the prose we have now, but we can supplement it with the autodoc generated documentation for the Python objects.

[1]: https://github.com/tomchristie/mkautodoc

Reply:
I expect if the links move around by a few lines people will still be able to find the class in question.  GitHub should do the fix, allowing something like `http.....thing.py?F-Class-NormalizeWhiteSpace` to search the blob for the first matching string.  

I'll go ahead and file a decent bug report on this.  The previous many times a bug was opened, there was no clear suggestion for GitHub to proceed.  GitHub is now getting the Microsoft mindset and yet-another-feature is likely to be looked on favorably.

Reply:
Hmm.... I'll still ask GitHub to put in the feature.  The people I know there left :(

When I'm done with this section, I'll also implement a post processor extension to allow links like `source:extensions/footnotes.py#F-Class-NormalizeWhiteSpace`, though it would use I/O access.

Reply:
How about I leave out the line references for now?  The examples will link to the correct file and an extension writer can scroll through the code.  A later PR could add fancy links and a post-processor for the doc building.

Reply:
That seems like a reasonable compromise.